### DISCUSSION TITLE
Dependencies with extensions

### DISCUSSION THREAD

**Discussion by maartenvn** [directus/directus#3671] [Feature Requests]
As of right now, dependencies in extensions (like interfaces, modules, ...) all require dependencies to be included inside the bundle of that extension.

### Problem with this approach

What if 2 extensions use the same dependency? Both dependencies will be included in the bundle of that extension, and loaded at runtime. Dependencies that inject anything in the global scope can be in conflict with each other. What if 2 interfaces are used in the same page and they both use the same dependency?

What do you think? Is there an alternative approach that could solve this issue?

**Comment by philleepflorence** [comment]
Is the approach the same or different in D8 and D9?

**Reply by maartenvn** [reply]
Directus 8 is doing the same thing, I think, because its also using rollup/parcel for bundling, which will include the dependencies.

**Reply by philleepflorence** [reply]
I assume it is also the same for custom extensions, correct?
Since there are item details that include multiple of the same interfaces, is this already happening?
Or do you have a scenario that could illustrate this more?

I am trying to recall if we talked about this a while back but had to keep extensions modular because of custom extension, or for another reason, maybe Rijk would remember and clarify.

**Comment by rijkvanzanten** [comment]
> What if 2 extensions use the same dependency? 

Then both include the same dependency. It increases the bundle size of each extension marginally because of the code duplication.

> Both dependencies will be included in the bundle of that extension, and loaded at runtime. Dependencies that inject anything in the global scope can be in conflict with each other. What if 2 interfaces are used in the same page and they both use the same dependency?

Dependencies really shouldn't attach anything into the global scope.. Otherwise—like you correctly state—you'll run into versioning issues

> What do you think?

I think it's an inevitability.. Seeing that the app is built before release, (afaik) there's no way to "expose" the same packages to the extensions. Likewise, extensions are built and distributed in isolation, so aren't really aware of each others existence 🤔 

> Is there an alternative approach that could solve this issue?

Theoretically speaking, extensions might be able to import extensions from "../../node_modules" and use the already installed ones in the app/api. Then on npm installs of extensions, we.. might.. ehh come up with a way where they can all inject dependencies into a shared..... package.json .. a couple directories up?? and then .. share..... i don't know 🤷🏻 

To be honest, besides the potential "global scope" issue (which I think should be fairly rare, as not a lot of things attach to `window` anymore), the downsides aren't that big. Sure, the extension bundle size is a little bigger than it has to be, but seeing that Directus is mostly an "app" (vs a website), initial load times aren't that big of a concern.

**Reply by philleepflorence** [reply]
Vue should be able to manage the loading and offloading of the component dependencies, if you leave an item details page and return - say when adding many new items, correct?
That way, memory should not be an issue.

**Reply by rijkvanzanten** [reply]
It's not so much about memory use once the component is on the page, it's about sharing node_modules between the extensions in build time. If you create lets say 2 (separate) custom modules that both use Bootstrap, you'll end up installing Bootstrap twice

**Reply by philleepflorence** [reply]
Oh sorry, I meant the dependencies, like Bootstrap, loaded as duplicates in memory (JS), or as references.
If they are duplicates, then would the whole component with dependencies would be loaded and offloaded by Vue?
I was thinking of JSON Objects while I was typing that.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
